### PR TITLE
Add telemetry for debug add breakpoints #55514

### LINKED ISSUE #55514: Telemetry: Add the ability to track usage of debug actions from editor gutter
When going over our telemetry for debug actions we have the ability to track usage counts of debug actions triggered from keyboard shortcuts and commands, as they execute commands which we emit telemetry for, but we don't seem to have telemetry in place for breakpoints set in the gutter (click or context menu). This is unfortunate as my hypothesis is that the majority of breakpoints are set this way.

https://github.com/Microsoft/vscode/blob/f82b14800cc0ed6f21a450b684518a4db05fc260/src/vs/workbench/parts/debug/browser/debugEditorActions.ts#L21

**Ask**
- Add ability to track breakpoint toggles (add/remove/change) from Gutter click
https://github.com/Microsoft/vscode/blob/bfc49dc41bcdbcaf0b64fe173415f6b9ce1ecd11/src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts#L222
- Add ability to track breakpoint toggles from gutter context menu 
https://github.com/Microsoft/vscode/blob/bfc49dc41bcdbcaf0b64fe173415f6b9ce1ecd11/src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts#L152

### PR DESCRIPTION
**Pull request by auchenberg**
Adds telemetry for debug add breakpoints by extending `addBreakpoints` in `debugService` with new `context` parameter. 

Emits new `debugAddBreakpoint` telemetry event with `hasCondition`, `hasHitCondition`,  `hasLogMessage`, and `context` properties.

Fixes #55514

### CODE REVIEW

File: src/vs/workbench/parts/debug/electron-browser/debugService.ts [hunk 1]
```diff
@@ -1064,4 +1065,27 @@ export class DebugService implements IDebugService {
...
+
+	private telemetryDebugAddBreakpoint(breakpoint: IBreakpoint, context: string): TPromise<any> {
+		/* __GDPR__
+			"debugAddBreakpoint" : {
+				"type" : { "classification": "SystemMetaData", "purpose": "FeatureInsight" },
+				"context": { "classification": "SystemMetaData", "purpose": "FeatureInsight" }
+			}
+		*/
+
+		var breakpointType = 'Breakpoint';
```
Comment (isidorn) [inline src/vs/workbench/parts/debug/electron-browser/debugService.ts]: Use `let` not `var`

Comment (isidorn) [reply src/vs/workbench/parts/debug/electron-browser/debugService.ts]: I do not like that we have this logic here: that we determing the type fo the breakpoint based on its characterstics. Do you really need this?
If yes I suggest to move this to be a field in the breakpoint called typeLabel and it shuold also be used in all places where we do this, for example [here](https://github.com/Microsoft/vscode/blob/6ba3f80ff402291d87e358624044d401a52e3547/src/vs/workbench/parts/debug/browser/breakpointsView.ts#L586)

File: src/vs/workbench/api/electron-browser/mainThreadDebugService.ts [hunk 2]
```diff
@@ -133,7 +133,7 @@ export class MainThreadDebugService implements MainThreadDebugServiceShape, IDeb
 						logMessage: l.logMessage
 					}
 				);
-				this.debugService.addBreakpoints(uri.revive(dto.uri), rawbps);
+				this.debugService.addBreakpoints(uri.revive(dto.uri), rawbps, 'mainThreadDebugService');
```
Comment (isidorn) [inline src/vs/workbench/api/electron-browser/mainThreadDebugService.ts]: This actually means that the breakpoint was added by an extension, so for readability I would potentially change the context

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@auchenberg Thanks for the PR, I have added comments in the code directly.
Aprat from that did you build vscode out of source and verify this gets called when you add a breakpoint just once?
How about if you add a breakpoint and start modifiying a file by updating it lines - it should onle be reported once.
Looking at the code and how I structured it that should be the case, but please verify by running out of source if you have not already. Thanks

File: src/vs/workbench/parts/debug/electron-browser/debugService.ts (hunk 1 above)
Comment (auchenberg) [reply src/vs/workbench/parts/debug/electron-browser/debugService.ts]: Yes we need this to be able to get an overview of the breakpoint types. Alternatively we can send the properties like `hasCondition`, `hasHitCondition` and `hasLogMessage` and do the aggregation as a part of the query? That way we move the logic our of VS Code core, and can easily re-categorize?

I wasn't aware of that functionality in the `breakpointsView` which really is render specific, but conceptually the same logic.

File: src/vs/workbench/api/electron-browser/mainThreadDebugService.ts (hunk 2 above)
Comment (auchenberg) [reply src/vs/workbench/api/electron-browser/mainThreadDebugService.ts]: Good point. Wasn't sure where this was used. Updated.

**Comment by auchenberg** [comment]
Yes I've validated that the telemetry events are flowing locally. Updating lines doesn't affect the addBreakpoint logic.

```
[2018-10-22 10:54:54.057] [telemetry] [trace] telemetry/debugAddBreakpoint {"properties":{"type":"Logpoint","context":"breakpointWidget","common.machineId":"6c9d2bc8f91b89624add29c0abeae7fb42bf539fa1cdb2e3e57cd668fa9bcead","sessionID":"428cfae4-e45b-423c-9c80-687cd877701c1540230829641","version":"1.29.0","common.platformVersion":"18.0.0","common.platform":"Mac","common.nodePlatform":"darwin","common.nodeArch":"x64",
```

**Comment by auchenberg** [comment]
I've push a changed so we now will send more "raw" data, and can move the breakpoint categorization to the aggregation side.

**Comment by isidorn** [comment]
@auchenberg this looks great now, thanks a lot. Merging in☀️

**Pull request merged**